tcl/generator-sv/sv.tm

```diff
@@ -0,0 +1,261 @@
+            set registerFile $cRegisterFile
+        }
+
+        public method produce {destinationPath {generator ""}} {
+            file mkdir $destinationPath
+            ::puts "svGenerator processing $registerFile > ${destinationPath}[$registerFile name].sv"
+            set res [produce_RegisterFile ]
+            odfi::files::writeToFile ${destinationPath}[$registerFile name].sv $res
+        }
+
+        public method getEnclosingRF {instance} {
+            if {[[$instance parent] isa osys::rfg::RegisterFile]} {
+                return [$instance parent]
+            } else {
+                return [getEnclosingRF [$instance parent]]
+            }
+        }
+
+        public method getGroupsName {instance} {
+            if {[[$instance parent] isa osys::rfg::Group]} {
+                if {[[$instance parent] isa osys::rfg::RegisterFile]} {
+                    return ""
+                } else {
+                    return "[getGroupsName [$instance parent]][[$instance parent] name]_"
+                }
+            } else {
+                return [getGroupsName [$instance parent]]
+            }
+        }
+
+        public method getFullName instance {
+            if {[$instance isa osys::rfg::RamBlock]} {
+                return "ram_[[getEnclosingRF $instance] name]_[getGroupsName $instance][$instance name]_c"
+            } elseif {[$instance isa osys::rfg::Register]} {
+                return "reg_[[getEnclosingRF $instance] name]_[getGroupsName $instance][$instance name]_c"
+            } elseif {[$instance isa osys::rfg::Group]} {
+                if {[$instance isa osys::rfg::RegisterFile]} {
+                    return "rf_[$instance name]_c"
+                } else {
+                    return "grp_[[getEnclosingRF $instance] name]_[getGroupsName $instance][$instance name]_c"
+                }
+            }
+        }
+
+        public method ld  x {
+            return [expr {int(ceil(log($x)/[expr log(2)]))}]
+        }
+
+        public method produce_RegisterFile args {
+            set out [odfi::common::newStringChannel]
+            set comp [list]
+            set comp_name [list]
+            $registerFile walkDepthFirst {
+                lappend comp $it
+                lappend comp_name [getFullName $it]
+                return true
+            }
+
+            foreach item $comp_name {
+                #-- while there are dublicates in comp
+                while {[llength [lsearch -all $comp_name $item]] > 1} {
+                    #-- delete the first dublicate
+                    set index [lsearch $comp_name $item]
+                    set comp [lreplace $comp $index $index]
+                    set comp_name [lreplace $comp_name $index $index]
+                }
+            }
+            #-- reverse list so classes are declared before use
+            set comp [lreverse $comp]
+
+            foreach item $comp {
+                if {[$item isa osys::rfg::RamBlock]} {
```

Comment:
There are odfi::common functions to create intends:
odfi::common::printlnIndent
odfi::common::printlnOutdent